### PR TITLE
Course Module

### PR DESCRIPTION
**Pull request by belish**
created course module and minor modification on Exception

please review @abrahamdoosti

### CODE REVIEW

File: src/main/java/com/example/spring/boot/exception/GenericExceptionMapper.java [hunk 1]
```diff
@@ -13,30 +16,32 @@
...
 	}
 
-   //common method for all exception categories.
-	private ResponseEntity<ExceptionEntity> handleException(String message, HttpStatus status) {
-		//LOGGER.error(message);
-		return new ResponseEntity<ExceptionEntity>(new ExceptionEntity(message,status.value(),"www.student.com"), status);
+	// common method for all exception categories.
+	private ResponseEntity<ExceptionEntity> handleException(String message, HttpStatus status, URI uri) {
+		// LOGGER.error(message);
+		return new ResponseEntity<ExceptionEntity>(new ExceptionEntity(message, status.value(), uri.getRawPath()),
```
Comment (belish) [inline src/main/java/com/example/spring/boot/exception/GenericExceptionMapper.java:43]: Found a way to pass uri

File: src/main/java/com/example/spring/boot/exception/GenericExceptionMapper.java [hunk 2]
```diff
@@ -13,30 +16,32 @@
 public class GenericExceptionMapper extends ResponseEntityExceptionHandler {
 
 	@ExceptionHandler(value = { Exception.class })
-	protected ResponseEntity<ExceptionEntity> handleUnkownException(Exception ex, WebRequest request) {
+	protected ResponseEntity<ExceptionEntity> handleUnkownException(Exception ex, WebRequest request) throws URISyntaxException {
 		// LOGGER.error(ex.getMessage(), ex);
-
-		return new ResponseEntity<ExceptionEntity>(new ExceptionEntity(ex.getMessage(), HttpStatus.INTERNAL_SERVER_ERROR.value(), "www.student.com"),
-				HttpStatus.INTERNAL_SERVER_ERROR);
+		return handleException(ex.getMessage(), HttpStatus.INTERNAL_SERVER_ERROR,
```
Comment (belish) [inline src/main/java/com/example/spring/boot/exception/GenericExceptionMapper.java:21]: Exception specific  uri

File: src/main/java/com/example/spring/boot/service/CourseServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+			throw new ResourceNotFoundException("No Courses found");
+		return new ResponseEntity<List<Course>>(courseDAO.getAllCourses(),HttpStatus.OK);
+	}
+
+	@Override
+	public ResponseEntity<Course> registerCourse(Course course) throws DuplicateResourceException {
+		if(!courseDAO.getAllCourses().contains(course)) {
+			return new ResponseEntity<Course>(courseDAO.addCourse(course), HttpStatus.CREATED);
+		}
+		throw new DuplicateResourceException("this Course is duplicate");
```
Comment (belish) [inline src/main/java/com/example/spring/boot/service/CourseServiceImpl.java:43]: have to change this to DuplicateResourceException(Course.class);

File: src/main/java/com/example/spring/boot/controller/StudentController.java [hunk 4]
```diff
@@ -33,27 +33,19 @@
...
 	}
 
-	@RequestMapping(value = "/", method = RequestMethod.POST, consumes = MediaType.APPLICATION_JSON_VALUE)
+	@RequestMapping(method = RequestMethod.POST, consumes = MediaType.APPLICATION_JSON_VALUE)
 	public ResponseEntity<Student> createStudent(@RequestBody final Student student) throws  DuplicateResourceException {
 		return studentService.registerStudent(student);
 	}
 	
 	@RequestMapping(value = "/{studentId}", method = RequestMethod.PUT, consumes = MediaType.APPLICATION_JSON_VALUE)
 	public ResponseEntity<Student> updateStudent(@RequestBody final Student student,@PathVariable("studentId") final int studentId) {
```
Comment (abrahamdoosti) [inline src/main/java/com/example/spring/boot/controller/StudentController.java]: student should not be final

**Review by abrahamdoosti — APPROVED** [review]
looks good

File: src/main/java/com/example/spring/boot/controller/StudentController.java (hunk 4 above)
Comment (belish) [reply src/main/java/com/example/spring/boot/controller/StudentController.java]: removed final

File: src/main/java/com/example/spring/boot/service/CourseServiceImpl.java (hunk 3 above)
Comment (belish) [reply src/main/java/com/example/spring/boot/service/CourseServiceImpl.java:43]: Changed

**Review by belish — COMMENTED** [review]
please review again, made changes

**Comment by abrahamdoosti** [comment]
👍

**Pull request merged**